components/places/src/storage/history_metadata.rs

```diff
@@ -0,0 +1,992 @@
+                get_host_and_port(:url),
+                reverse_host(get_host_and_port(:url)),
+                -1
+            )",
+                &[(":url", &url.as_str())],
+            )?;
+            Some(tx.conn().last_insert_rowid())
+        }
+    };
+
+    let sql = "INSERT INTO moz_places_metadata
+        (guid, place_id, created_at, updated_at, total_view_time, search_query_id, is_media, parent_domain_id)
+    VALUES
+        (:guid, :place_id, :created_at, :updated_at, :total_view_time, :search_query_id, :is_media, :parent_domain_id)";
+
+    let guid = SyncGuid::random();
+    tx.execute_named_cached(
+        sql,
+        &[
+            (":guid", &guid),
+            (":place_id", &places_id),
+            (":created_at", &metadata.created_at), // we probably just want to auto-generate these.
+            (":updated_at", &metadata.updated_at),
+            (":total_view_time", &metadata.total_view_time),
+            (":search_query_id", &search_query_id),
+            (":is_media", &metadata.is_media),
+            (":parent_domain_id", &parent_domain_id),
+        ],
+    )?;
+    Ok(guid)
+}
+
+pub fn add_metadata(db: &PlacesDb, metadata: HistoryMetadata) -> Result<SyncGuid> {
+    let places_id = db.try_query_one(
+        "SELECT id FROM moz_places WHERE url_hash = hash(:url) AND url = :url",
+        &[(":url", &metadata.url)],
+        true,
+    )?;
+
+    // Look up the search query first, maybe it's already stored in the db.
+    // Do this before starting a transaction, since it's a SELECT and doesn't need to be part of a tx.
+    // We depend on having a single write connection, so we know a search term won't be inserted by another caller.
+    // NB: there is also a sync writer, but we don't currently sync any of this data.
+    let search_query = match &metadata.search_term {
+        None => SearchQueryEntry::None,
+        Some(term) => {
+            let lowercase_term = term.to_lowercase();
+            match db.try_query_one(
+                "SELECT id FROM moz_places_metadata_search_queries WHERE term = :term",
+                &[(":term", &lowercase_term)],
+                true,
+            )? {
+                Some(id) => SearchQueryEntry::Existing(id),
+                None => SearchQueryEntry::CreateFor(lowercase_term),
+            }
+        }
+    };
+
+    let parent_domain = match &metadata.parent_url {
+        None => ParentDomainEntry::None,
+        Some(parent_url) => {
+            let parent_url = Url::parse(parent_url)?;
+            match db.try_query_one(
+                "SELECT id FROM moz_origins WHERE prefix = get_prefix(:url) AND host = get_host_and_port(:url)",
+                &[(":url", &parent_url.as_str())],
+                true
+            )? {
+                Some(id) => ParentDomainEntry::Existing(id),
+                None => ParentDomainEntry::CreateFor(parent_url)
+            }
+        }
+    };
+
+    let tx = db.begin_transaction()?;
```

Comment:
as implied above, I think we probably want this at the top of the function, just to ensure nothing races with us (in theory sync could - it uses its own write connection.)